Add weak pointer and finalizer support directly to THStorage.

ezyang opened the pull request:
The underlying use-case is the file descriptor to storage cache in
torch.multiprocessing.reductions.  Previously, this was implemented by wrapping
an existing allocator with a "weak ref" allocator which also knew to null out
the weak reference when the storage died.  This is terribly oblique, and
prevents us from refactoring the allocators to get rid of per-storage allocator
state.

So instead of going through this fiasco, we instead directly implement weak
pointers and finalizers in THStorage.  Weak pointers to THStorage retain the
THStorage struct, but not the data_ptr.  When all strong references die,
data_ptr dies and the finalizers get invoked.

There is one major hazard in this patch, which is what happens if you
repeatedly call _weak_ref on a storage.  For cleanliness, we no longer
shove our grubby fingers into the finalizer struct to see if there is already
a Python object for the weak reference and return it; we just create a new one
(no one is checking these Python objects for identity).  This means if you
keep calling it, we'll keep piling on finalizers.  That's bad! But I am
not going to fix it until it is actually a problem for someone, because
then we need to add another caching layer.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Add weak pointer and finalizer support directly to THStorage.

The underlying use-case is the file descriptor to storage cache in
torch.multiprocessing.reductions.  Previously, this was implemented by wrapping
an existing allocator with a "weak ref" allocator which also knew to null out
the weak reference when the storage died.  This is terribly oblique, and
prevents us from refactoring the allocators to get rid of per-storage allocator
state.

So instead of going through this fiasco, we instead directly implement weak
pointers and finalizers in THStorage.  Weak pointers to THStorage retain the
THStorage struct, but not the data_ptr.  When all strong references die,
data_ptr dies and the finalizers get invoked.

There is one major hazard in this patch, which is what happens if you
repeatedly call _weak_ref on a storage.  For cleanliness, we no longer
shove our grubby fingers into the finalizer struct to see if there is already
a Python object for the weak reference and return it; we just create a new one
(no one is checking these Python objects for identity).  This means if you
keep calling it, we'll keep piling on finalizers.  That's bad! But I am
not going to fix it until it is actually a problem for someone, because
then we need to add another caching layer.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

colesbury commented on torch/csrc/generic/StorageSharing.cpp:
```diff
@@ -418,8 +407,9 @@ PyObject * THPStorage_(isShared)(THPStorage *self)
-      allocator == &THManagedSharedAllocator) {
+      allocator == &THManagedSharedAllocator ||
+      // This is a bit racy...
+      self->cdata->weakcount.load() > 1) {
```
This line worries me. Is it necessary? Won't every shared Storage now have one of the above allocators?

ezyang replied:
Ah! If it's true that the only storages that are allowed to `weak_ref` are these two, we can drop this line.

I didn't see anything that guaranteed this in this C++ file, but maybe it's by convention globally?

colesbury commented on aten/src/THC/THCStorage.cpp:
```diff
@@ -60,25 +62,25 @@ THCStorage* THCStorage_newWithAllocator(THCState *state,
-    }
-    if(self->flag & TH_STORAGE_VIEW) {
-      THCStorage_free(state, self->view);
+  if ((storage->flag & TH_STORAGE_REFCOUNTED) && (storage->refcount.load() > 0)) {
```
The `storage->refcount.load() > 0` checks feels wrong. If we're going to detect that we're in an inconsistent state, should throw an error or print a warning. 

I know this is from the THStorage_free version, so we can deal with it at some later point.

ezyang replied:
Yeah, we should drop it. Also we should drop retainIfLive, it doesn't really have a place in this new world order.

colesbury commented on aten/src/THC/THCStorage.cpp:
```diff
@@ -174,7 +176,9 @@ THCStorage* THCStorage_newWithDataAndAllocator(
   storage->data_ptr = data;
   storage->size = size;
-  storage->refcount = 1;
+  new (&storage->refcount) std::atomic<int>(1);
```
I wonder if it's worth adding a real constructor to THStorage and calling that.

ezyang replied:
Yes we should :)

colesbury commented on torch/multiprocessing/reductions.py:
```diff
@@ -22,6 +22,9 @@ class StorageRef(object):
     def __init__(self, ptr):
         self.cdata = ptr
 
+    def __del__(self):
```
In the above comment, I don't think anything sets "cdata" to None anymore.

If I understand correctly:

- `shared_cache` has a weak reference to a `StorageRef`
- `StorageRef` morally holds a weak reference to a THStorage pointer
- `THStorage` holds a strong reference to `StorageRef` via finalizer

So when all the THStorage strong refs go away:

- The finalizer runs (does nothing)
- The finalizer gets destructed (decrements StorageRef's ref count)
- StorageRef drops a weak reference of the THStorage
- This may lead to 0 weak refs on THStorage which frees the THStorage struct

It would be nice to simplify this, but I don't see an easy way to do it. It would be nice if a weak ref to a `torch.Storage` but a weakref to a `torch.Storage` Python object is not the same as a weak ref to a `THStorage`. (You could make it so by making `THStorage` own it's wrapper).

ezyang replied:
Your understanding is right.

> You could make it so by making THStorage own it's wrapper

Right, and it would be very similar to what we do for Tensor today (to preserve `id()` across FFI crossings). I didn't do this, mostly because I saw a way to do it with the finalizer instead. Note that you are always going to need a finalizer, because the destructor for THStorage isn't going to know how to decref a Python object, so someone else has to give it a hand. If you want me to do it this way, I can; it's not too much more work.

colesbury commented on torch/csrc/finalizer.h:
```diff
@@ -0,0 +1,23 @@
+namespace torch {
+
+struct PyObjectFinalizer : public THFinalizer {
+  THPObjectPtr pyobj_;
```
This looks like it will call decrement the Python reference count in the destructor, which might not be holding the GIL?

ezyang replied:
You're right, this is a bug.

ezyang replied:
Isn't this a hella big hazard with THPObjectPtr? Shouldn't we make its destructor automatically take out the GIL?

colesbury replied:
Yeah, it's a hazard, but most use case are on the stack (i.e. implicitly thread-local), so I'm not sure it's worth the perf hit.

colesbury reviewed (commented):
Generally looks good.

I think the THPObjectPtr destruction in PyObjectFinalizer without the GIL is a bug, but I'm not sure.

ezyang committed:
CR comments

- Don't consider storages with weak references as "shared" (previously, they
  were unconditionally considered so.)
- Remove unnecessary "defensive" check that we don't double-free storage
- Delete retainIfLive, it is dead
- GIL-protect deallocation of THPObjectPtr

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

apaszke commented on torch/multiprocessing/reductions.py:
```diff
@@ -127,6 +130,10 @@ def reduce_storage(storage):
+    # WARNING!  This call to _weak_ref could lead to O(n) deleter
+    # behavior, if you repeatedly call it on the same Storage (all
+    # other sites are guarded by shared_cache; maybe this site
+    # should be too?)
```
This warning applies to multiple places in this file I think

ezyang replied:
Actually, it doesn't, because in all the other cases, there's a test if it's already in `shared_cache`, as the comment says :)

apaszke reviewed (commented):
Well the O(n) run time is not too bad, since it's still O(1) if you amortize it, and I'm more worried about "space leaks" due to unbounded length of finalizer chains.

ezyang commented on aten/src/TH/THStorage.hpp:
```diff
@@ -10,16 +10,43 @@
+//      plus one more if refcount > 0
+//
+//  - THStorage stays live as long as there are any strong
+//    or weak pointers to it (refcount > 0 || weakcount > 0)
```
Yes.

ezyang commented on aten/src/TH/THStorage.hpp:
```diff
@@ -10,16 +10,43 @@
+//
+//  - finalizers are called and data_ptr is deallocated when refcount == 0
+//
+//  - Once refcount == 0, it can never again be > 0 (it's monotonic)
```
Yes.

ezyang committed:
CR comments

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please
